Fix document error. (#611) (#617)
crystal-lang/crystal-book: docs/guides/concurrency.md

At the moment of this writing, Crystal has concurrency support but not parallelism: several tasks can be executed, and a bit of time will be spent on each of these, but two code paths are never executed at the same exact time.

> Crystal is still single-threaded by default, except for the garbage collector (currently [Boehm GC](http://www.hboehm.info/gc/)). Parallelism is supported, but it is currently considered experimental. Check out [this Crystal Blog post about parallelism](https://crystal-lang.org/2019/09/06/parallelism-in-crystal.html) for more information.

Reviewer: IMO the previous wording about "Crystal program" is better.

```suggestion
A Crystal program by default executes in a single operating system thread, except for the garbage collector (currently [Boehm GC](http://www.hboehm.info/gc/)). Parallelism is supported, but it is currently considered experimental. Check out [this Crystal Blog post about parallelism](https://crystal-lang.org/2019/09/06/parallelism-in-crystal.html) for more information.
```

Author: I saw you remove `mark-and-sweep` in document, so, that description is not accurate, right?

Author: Okay, maybe we can have a separate section, do a little more detail explain about GC.

> Crystal is still single threaded by default. Parallelism is supported, but is currently considered experimental. Checkout [Parallelism in Crystal - The Crystal Programming Language for more information](https://crystal-lang.org/2019/09/06/parallelism-in-crystal.html), only one exception, the Garbage Collector (GC) which implements a concurrent mark-and-sweep (currently [Boehm GC](http://www.hboehm.info/gc/)).

Reviewer 2: I think this would be more clear for English readers if it were
```md
Crystal is still single-threaded by default, except for the garbage collector (currently [Boehm GC](http://www.hboehm.info/gc/)).
Parallelism is supported, but it is currently considered experimental.
Check out [the Crystal Blog post about parallelism](https://crystal-lang.org/2019/09/06/parallelism-in-crystal.html) for more information.
```
The  markdown formatting doesn't matter to me.

> Crystal is still single threaded by default. Parallelism is supported, but is currently considered experimental. Checkout [Parallelism in Crystal - The Crystal Programming Language for more information](https://crystal-lang.org/2019/09/06/parallelism-in-crystal.html), only one exception, the Garbage Collector (GC) which implements a concurrent mark-and-sweep (currently [Boehm GC](http://www.hboehm.info/gc/)).

Reviewer 2: ```suggestion
Crystal is still single-threaded by default, except for the garbage collector (currently [Boehm GC](http://www.hboehm.info/gc/)). Parallelism is supported, but it is currently considered experimental. Check out [this Crystal Blog post about parallelism](https://crystal-lang.org/2019/09/06/parallelism-in-crystal.html) for more information.
```

```
Before receive
Before send
After send
After receive
```

> First, the program spawns a fiber but doesn't execute it yet. When we invoke `channel.receive`, the main fiber blocks and execution continues with the spawned fiber. Then `channel.send(nil)` is invoked, Note that this `send` does not occupy space in the channel, This is because there is a `receive` invoked prior to the first `send`, so this `send` not got blocked. Fiber only switch out when blocked or executing to completion, so, will continue the codes after `send`, switch back to main fiber when finish `puts "After send"`.

Reviewer 2: ```suggestion
First, the program spawns a fiber but doesn't execute it yet. When we invoke `channel.receive`, the main fiber blocks and execution continues with the spawned fiber. Then `channel.send(nil)` is invoked. Note that this `send` does not occupy space in the channel because there is a `receive` invoked prior to the first `send`, `send` is not blocked. Fibers only switch out when blocked or executing to completion. So the spawned fiber will continue after `send`, and execution will switch back to main fiber once `puts "After send"` is executed.
```

> Main fiber then resume at `channel.receive`, which was waiting for a value. Then the main fiber continues executing and finishes.

Reviewer: ```suggestion
The main fiber then resumes at `channel.receive`, which was waiting for a value. Then the main fiber continues executing and finishes.
```

```
Before first receive
Before first send
Before second send
1
Before second receive
2

```

> Note that when the program executes a `receive`, that fiber blocks and execution continues with the other fiber. When `channel.send(1)` is executed, execution continues because non-blocked, until `channel.send(2)`, because channel blocked, execution continues with the fiber that was waiting on that channel.

Reviewer 2: ```suggestion
Note that when the program executes a `receive`, the current fiber blocks and execution continues with the other fiber. When `channel.send(1)` is executed, execution continues because `send` is non-blocking if the channel is not yet full. However, `channel.send(2)` does cause the fiber to block because the channel (which has a size of 1 by default) is full, so execution continues with the fiber that was waiting on that channel.
```
